python/pyspark/ml/torch/distributor.py

```diff
@@ -0,0 +1,491 @@
+#       once we are in agreement about which functions should be in utils.py
+def get_conf_boolean(sc: SparkContext, key: str, default_value: str) -> bool:
+    """Get the conf "key" from the given spark context,
+    or return the default value if the conf is not set.
+    This expects the conf value to be a boolean or string;
+    if the value is a string, this checks for all capitalization
+    patterns of "true" and "false" to match Scala.
+
+    Parameters
+    ----------
+    sc : SparkContext
+        The SparkContext for the distributor.
+    key : str
+        string for conf name
+    default_value : str
+        default value for the conf value for the given key
+
+    Returns
+    -------
+    bool
+        Returns the boolean value that corresponds to the conf
+
+    Raises
+    ------
+    Exception
+        Thrown when the conf value is not a boolean
+    """
+    val = sc.getConf().get(key, default_value)
+    lowercase_val = val.lower()
+    if lowercase_val == "true":
+        return True
+    if lowercase_val == "false":
+        return False
+    raise Exception(
+        "get_conf_boolean expected a boolean conf "
+        "value but found value of type {} "
+        "with value: {}".format(type(val), val)
+    )
+
+
+def get_gpus_owned(addresses: list[str]) -> list[str]:
+    """
+    Gets the number of GPUs that Spark scheduled to the calling task.
+    Returns:
+        The number of GPUs that Spark scheduled to the calling task.
+    """
+    CUDA_VISIBLE_DEVICES = "CUDA_VISIBLE_DEVICES"
+    pattern = re.compile("^[1-9][0-9]*|0$")
+    if any(not pattern.match(address) for address in addresses):
+        raise ValueError(
+            f"Found GPU addresses {addresses} which "
+            "are not all in the correct format "
+            "for CUDA_VISIBLE_DEVICES, which requires "
+            "integers with no zero padding."
+        )
+    if CUDA_VISIBLE_DEVICES in os.environ:
+        gpu_indices = list(map(int, addresses))
+        gpu_list = os.environ[CUDA_VISIBLE_DEVICES].split(",")
+        gpu_owned = [gpu_list[i] for i in gpu_indices]
+        return gpu_owned
+    return addresses
+
+
+def create_torchrun_command(input_params: dict[str, Any], train_path: str, *args: Any) -> list[str]:
+    """Returns the expected torchrun command based on the input.
+
+    Parameters
+    ----------
+    input_params : dict[str, Any]
+        The dictionary of the input parameters of the distributor. The most relevant params
+        are local_mode and num_processes.
+    train_path : str
+        The path to the (potentially autogenerated) train.py file
+    args: *args
+        The input arguments to the train.py file.
+
+    Returns
+    -------
+    str
+        The output torchrun command
+    """
+    local_mode = input_params["local_mode"]
+    num_processes = input_params["num_processes"]
+
+    if local_mode:
+        standalone = ["--standalone", "--nnodes=1"]
```

Comment:
variable name `standalone` is not clear.
use `torchrun_options` instead

Reply:
Sure